cubedash/summary/_extents.py

```diff
@@ -622,6 +622,37 @@ def datasets_by_region(
     )
 
 
+def products_by_region(
+    engine: Engine,
+    index: Index,
+    region_code: str,
+    time_range: Range,
+    limit: int,
+    offset: int = 0,
+) -> Generator[Dataset, None, None]:
+    query = (
+        select(postgres_api._DATASET_SELECT_FIELDS)
+        .select_from(
+            DATASET_SPATIAL.join(DATASET, DATASET_SPATIAL.c.id == DATASET.c.id)
+        )
+        .where(DATASET_SPATIAL.c.region_code == bindparam("region_code", region_code))
+    )
+    if time_range:
+        query = query.where(
+            DATASET_SPATIAL.c.center_time > bindparam("from_time", time_range.begin)
+        ).where(DATASET_SPATIAL.c.center_time < bindparam("to_time", time_range.end))
+    query = (
+        query.order_by(DATASET_SPATIAL.c.center_time)
+        .limit(bindparam("limit", limit))
+        .offset(bindparam("offset", offset))
+    )
+
+    return (
+        res.dataset_type_ref
+        for res in engine.execute(query).fetchall()
+    )
```

Comment:
This query will be _very_ slow on our main instances. And it returns one-row-per-dataset rather than one-per-product, which might be a mistake? (the name says "find product". And when it hits the row limit from datasets, will be missing products in its return value)

Example improvement:

```suggestion
) -> Generator[DatasetType, None, None]:
    query = (
        select([DATASET_SPATIAL.c.dataset_type_ref]).distinct()
        .where(DATASET_SPATIAL.c.region_code == bindparam("region_code", region_code))
    )
    if time_range:
        query = query.where(
            DATASET_SPATIAL.c.center_time > bindparam("from_time", time_range.begin)
        ).where(DATASET_SPATIAL.c.center_time < bindparam("to_time", time_range.end))
    
    query = (
        query.order_by(DATASET_SPATIAL.c.dataset_type_ref)
        .limit(bindparam("limit", limit))
        .offset(bindparam("offset", offset))
    )

    return (
        index.products.get_by_name(res.dataset_type_ref)
        for res in engine.execute(query).fetchall()
    )
```

Ie.
- Return distinct products, rather than the product repeated for every dataset. 
- Remove the join (since we're throwing away all dataset information anyway)
- Only retrieve the `dataset_type_id` (so we don't need to transfer so much from the db)
- Return a DatasetType object (ie. ODC's Product), since it's called `find_product()`.

... we'll need to update the other code to expect `DatasetType.id` rather than plain id, though.

---

... but this still may not be fast enough on our bigger instances? We could add a dedicated index for it, possibly:

Eg.
```
$ select distinct dataset_type_ref from cubedash.dataset_spatial where region_code='53JMN' limit 100;
 dataset_type_ref 
══════════════════
               90
               91
(2 rows)

Time: 212.347 ms
$ create index region_code_products on cubedash.dataset_spatial(region_code, dataset_type_ref);
$ select distinct dataset_type_ref from cubedash.dataset_spatial where region_code='53JMN' limit 100;
 dataset_type_ref 
══════════════════
               90
               91
(2 rows)

Time: 16.291 ms
```
(was after warmup)